NPE in ClassDumper

After adding the the linkage-checker-enforcer-rules 1.2.0 to my pom, with:

```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-enforcer-plugin</artifactId>
                <version>3.0.0-M3</version>
                <dependencies>
                    <dependency>
                        <groupId>com.google.cloud.tools</groupId>
                        <artifactId>linkage-checker-enforcer-rules</artifactId>
                        <version>1.2.0</version>
                    </dependency>
                </dependencies>
                <executions>
                    <execution>
                        <id>enforce-linkage-checker</id>
                        <phase>verify</phase>
                        <goals>
                            <goal>enforce</goal>
                        </goals>
                        <configuration>
                            <rules>
                                <LinkageCheckerRule implementation= "com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule">
                                    <level>WARN</level>
                                </LinkageCheckerRule>
                            </rules>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

I encountered the following exception, immediately after the enforcer plugin gave the following warnings

```
[WARNING] The POM for xerces:xerces-impl:jar:2.6.2 is missing, no dependency information available
[WARNING] The POM for net.sf.ehcache:sizeof-agent:jar:1.0.1 is missing, no dependency information available
[WARNING] The POM for com.ibm.websphere:websphere_uow_api:jar:0.0.1 is missing, no dependency information available
[WARNING] The POM for com.ibm.websphere:uow:jar:6.0.2.17 is missing, no dependency information available
[WARNING] net.sf.ehcache:sizeof-agent:jar:1.0.1 was not resolved. Dependency path: com.xyz:internal-util:jar:1-SNAPSHOT > com.xyz:internal-shaded:jar:1.234 (compile) > net.sf.ehcache:ehcache-core:jar:2.6.11 (compile) > net.sf.ehcache:sizeof-agent:jar:1.0.1 (provided)
```


```
Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.NullPointerException
    at org.apache.bcel.generic.MethodGen.<init> (MethodGen.java:182)
    at com.google.cloud.tools.opensource.classpath.ClassDumper.catchesLinkageError (ClassDumper.java:500)
    at com.google.cloud.tools.opensource.classpath.ClassDumper.catchesLinkageError (ClassDumper.java:518)
    at com.google.cloud.tools.opensource.classpath.LinkageChecker.findSymbolProblem (LinkageChecker.java:482)
    at com.google.cloud.tools.opensource.classpath.LinkageChecker.lambda$findSymbolProblems$1 (LinkageChecker.java:180)
    at com.google.common.collect.ImmutableMultimap.lambda$null$2 (ImmutableMultimap.java:606)
    at java.lang.Iterable.forEach (Iterable.java:75)
    at com.google.common.collect.ImmutableMultimap.lambda$forEach$3 (ImmutableMultimap.java:606)
    at com.google.common.collect.RegularImmutableMap.forEach (RegularImmutableMap.java:185)
    at com.google.common.collect.ImmutableMultimap.forEach (ImmutableMultimap.java:605)
    at com.google.cloud.tools.opensource.classpath.LinkageChecker.findSymbolProblems (LinkageChecker.java:150)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.execute (LinkageCheckerRule.java:203)
    at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:205)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

It seems there was unexpected class file somewhere in the linkage checker’s class path.


```
Caused by: java.lang.NullPointerException
    at org.apache.bcel.generic.MethodGen.<init> (MethodGen.java:182)
    at com.google.cloud.tools.opensource.classpath.ClassDumper.catchesLinkageError (ClassDumper.java:500)
    at com.google.cloud.tools.opensource.classpath.ClassDumper.catchesLinkageError (ClassDumper.java:518)

```

Would you share a minimum pom.xml that produces the error? I’m interested in dependencies and dependencyManagement elements if any.

Thank you for the pom.xml. I'm able to reproduce the error. Continuing investigation.

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce (enforce-linkage-checker) on project linkage-checker-error: Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed.: NullPointerException -> [Help 1]
```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -481,7 +481,9 @@ class ClassDumper {
 
       for (Method method : sourceJavaClass.getMethods()) {
         if (method.getCode() == null) {
-          // No need to check the presence of try-catch clause for methods without code
+          // No need to check the presence of try-catch clause for methods without code. This guard
+          // avoids NullPointerException by BCEL.
+          // https://issues.apache.org/jira/browse/BCEL-336
           continue;
         }
         MethodGen methodGen = new MethodGen(method, sourceClassName, classGen.getConstantPool());

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -480,6 +480,10 @@ class ClassDumper {
       ClassGen classGen = new ClassGen(sourceJavaClass);
 
       for (Method method : sourceJavaClass.getMethods()) {
+        if (method.getCode() == null) {
+          // If the method doesn't have body, no need to check try-catch clause.
+          continue;
+        }
         MethodGen methodGen = new MethodGen(method, sourceClassName, classGen.getConstantPool());
         CodeExceptionGen[] exceptionHandlers = methodGen.getExceptionHandlers();
         for (CodeExceptionGen codeExceptionGen : exceptionHandlers) {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -481,7 +481,7 @@ class ClassDumper {
 
       for (Method method : sourceJavaClass.getMethods()) {
         if (method.getCode() == null) {
-          // If the method doesn't have body, no need to check try-catch clause.
+          // No need to check the presence of try-catch clause for methods without code
           continue;
         }
         MethodGen methodGen = new MethodGen(method, sourceClassName, classGen.getConstantPool());

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
@@ -405,4 +405,15 @@ public class ClassDumperTest {
                 (ClassFile classFile) -> classFile.getBinaryName(), "has class name"))
         .contains("com.ibm.icu.util.DateRule");
   }
+
+  @Test
+  public void testMethodCodeNull() throws IOException {
+    // catchesLinkageError should not raise an exception for MailDateFormat's constructor's null
+    // code.
+    // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1352
+    List<ClassPathEntry> classPath = resolvePaths("javax:javaee-api:6.0");
+    ClassDumper classDumper = ClassDumper.create(classPath);
+
+    classDumper.catchesLinkageError("javax.mail.internet.MailDateFormat");
+  }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
@@ -414,6 +414,8 @@ public class ClassDumperTest {
     List<ClassPathEntry> classPath = resolvePaths("javax:javaee-api:6.0");
     ClassDumper classDumper = ClassDumper.create(classPath);
 
-    classDumper.catchesLinkageError("javax.mail.internet.MailDateFormat");
+    // This should not raise NullPointerException
+    boolean result = classDumper.catchesLinkageError("javax.mail.internet.MailDateFormat");
+    assertFalse(result);
   }
 }

